external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -383,23 +403,24 @@ private void commitOffsetsForAckedTuples() {
     @Override
     public void ack(Object messageId) {
         final KafkaSpoutMessageId msgId = (KafkaSpoutMessageId) messageId;
-        if(!emitted.contains(msgId)) {
-            LOG.debug("Received ack for tuple this spout is no longer tracking. Partitions may have been reassigned. Ignoring message [{}]", msgId);
-            return;
-        }
-        
-        if (!consumerAutoCommitMode) {  // Only need to keep track of acked tuples if commits are not done automatically
-            acked.get(msgId.getTopicPartition()).add(msgId);
+        if (!emitted.contains(msgId)) {
+            LOG.debug("Received ack for message [{}], which is associated with either a null tuple that was never emitted, " +
```

Comment:
This message is ambiguous now. Could you add the msgId to emitted before acking (i.e. basically fake a "real" ack) instead? If people want to log that they're dropping an invalid tuple they can do it from the RecordTranslator without ambiguity.